### PR TITLE
pkg/daemon: run main loop

### PR DESCRIPTION
**Pull request by jlebon**
I may be mistaken, but it seems like what we want here is a main loop
that continuously watches for updates and re-applies config changes.
I.e., we don't want to reboot if a config change doesn't actually
requires it.

Of course, this is a moot point for now given that we always reboot for
any changes. Though figuring out the higher-level logic upfront is worth
it.

### CODE REVIEW

**Comment by ashcrow** [comment]
> Of course, this is a moot point for now given that we always reboot for
any changes. Though figuring out the higher-level logic upfront is worth
it.

I don't disagree and thank you for doing this PR! Note that ` daemon.Run` takes in a stop channel but does not use it. I believe it was added with the idea that one could write to the channel to stop the loop. I think it would make the most sense to have the loop within `daemon.Run` and use the stop channel as the method for stopping the run loop.

**Comment by ashcrow** [comment]
@jlebon that's fair as long as we keep to the top level thread as noted.

This is looking good. Let me know when you want a deeper review!

**Comment by sdemos** [comment]
I'm totally on board with this change. I've been thinking that the top-level logic needs a good refactor to be clearer.

It looks like with the changes the `triggerUpdate` function will be called everytime, so it seems like that will need to have a change in behavior. Right now the daemon will only set the update as done if `triggerUpdate` returns, which as you point out it will never do, so there needs to be another mechanism for setting updates as done.

**Comment by jlebon** [comment]
OK, updated this! Tested working, though it depends on #49.

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -65,20 +71,27 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) error {
...
 	if err := dn.writeUnits(newConfig.Spec.Config.Systemd.Units); err != nil {
-		return err
+		return false, err
 	}
 
 	dn.deleteStaleData(oldConfig, newConfig)
 
-	return nil
+	// gross simplification: if only files were touched and no units, then don't
+	// trigger a reboot
```
[the remark this answers is not in the document]
Comment (sdemos) [reply pkg/daemon/update.go]: the other example I had in mind was updating configuration files in `/etc`. there will be some programs that pick that up automatically, some that need to be told to reup the config, and some that need to be restarted, and there's not really a good way to tell which one is which, hence the full system reboot.

[the remark this answers is not in the document]
Comment (ashcrow) [reply pkg/daemon/update.go]: Agreed. I think for the time being we should always reboot with the idea that we build up a whitelist and implement it as a optimization task.

File: pkg/daemon/update.go [hunk 2]
```diff
@@ -286,19 +299,30 @@ func (dn *Daemon) writeFiles(files []ignv2_2types.File) error {
 	return nil
 }
 
+// This is essentially ResolveNodeUidAndGid() from Ignition; XXX should dedupe
```
Comment (ashcrow) [inline pkg/daemon/update.go]: Did https://github.com/openshift/machine-config-operator/pull/49 sneak in? :smile:

File: pkg/daemon/update.go [hunk 3]
```diff
@@ -65,20 +71,27 @@ func (dn *Daemon) update(oldConfig, newConfig *mcfgv1.MachineConfig) error {
 // whatever has been written is picked up by the appropriate daemons, if
 // required. in particular, a daemon-reload and restart for any unit files
 // touched.
-func (dn *Daemon) updateFiles(oldConfig, newConfig *mcfgv1.MachineConfig) error {
+func (dn *Daemon) updateFiles(oldConfig, newConfig *mcfgv1.MachineConfig) (bool, error) {
```
Comment (ashcrow) [inline pkg/daemon/update.go]: nit: I think a comment noting what the `bool` in the return would be helpful when reading the code.

**Pull request merged**